<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>Anti-malarial SNPs by Setfelix</title>
  </head>

  <body>
    <header>
      <div class="inner">
        <h1>Anti-malarial SNPs</h1>
        <h2>Pipeline to detect SNPs in specific codons of genes associated with antimalarial drug resistance</h2>
        <a href="https://github.com/Setfelix/anti_malaria_snps" class="button"><small>View project on</small> GitHub</a>
      </div>
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
          <h1>
<a id="detect-snps-in-codons-of-antimalarial-drug-resistance-genes" class="anchor" href="#detect-snps-in-codons-of-antimalarial-drug-resistance-genes" aria-hidden="true"><span class="octicon octicon-link"></span></a>Detect SNPs in codons of antimalarial drug resistance genes</h1>

<p><code>anti_malaria_snps.sh</code> is a BASH-scripted pipeline that can be used to identify SNPs in specific codons of 
<em>Plasmodium falciparum</em> genes that are associated with antimalarial drug resistance. 
Currently, the following genes and codons are supported:</p>

<table>
<thead>
<tr>
<th>Gene</th>
<th align="center">Codons</th>
</tr>
</thead>
<tbody>
<tr>
<td><em>pfmdr1</em></td>
<td align="center">86, 184, 1034, 1042,1246</td>
</tr>
<tr>
<td><em>pfcrt</em></td>
<td align="center">72, 73, 74, 75, 76</td>
</tr>
<tr>
<td><em>dhps</em></td>
<td align="center">436, 437, 540, 581, 613</td>
</tr>
<tr>
<td><em>dhfr</em></td>
<td align="center">51, 59, 108, 164</td>
</tr>
</tbody>
</table>

<p>Using ABI Sanger sequencing trace files of whole gene (or regions spanning codons of interest) from several samples, 
this pipeline outputs a CSV-formatted file of sample name, codons and their corresponding amino acids.</p>

<p>This pipeline is a combination of existing tools (such as <a href="http://bio-bwa.sourceforge.net/">BWA</a>, <a href="http://sourceforge.net/projects/sam2fasta/files/">sam2fasta.py</a>, <a href="https://github.com/bow/abifpy">abifpy</a>) and custom scripts and is made up of the following steps:</p>

<ol>
<li>Base calling and quality control, </li>
<li>Alignment of trimmed and filtered sequences to coding sequence of respective reference gene, </li>
<li>Conversion of alignment from SAM to FASTA format, </li>
<li>Parse FASTA alignment to output codons and corresponding amino acid in CSV format.</li>
</ol>

<hr>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<ol>
<li>First, ensure that the dependencies are installed. Follow the respective links to install: <a href="http://bio-bwa.sourceforge.net/">BWA</a>, <a href="http://sourceforge.net/projects/sam2fasta/files/">sam2fasta.py</a>, <a href="https://github.com/bow/abifpy">abifpy</a>
</li>
<li>Download the source <a href="https://github.com/Setfelix/anti_malaria_snps/zipball/master"><code>.zip</code></a> or <a href="https://github.com/Setfelix/anti_malaria_snps/tarball/master"><code>.tar.gz</code></a> file and extract</li>
<li>Optionally, add the anti_malaria_snps directory to your <code>$PATH</code> (in <code>.bashrc</code>, for example, to make it persistent) </li>
</ol>

<hr>

<h2>
<a id="usage" class="anchor" href="#usage" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage</h2>

<pre><code>     bash anti_malaria_snps.sh &lt;directory_of_ab1_files&gt; &lt;reference_gene_coding_sequence&gt; &lt;gene&gt; &lt;quality_cutoff&gt;
</code></pre>

<p>where: </p>

<pre><code>    &lt;directory_of_ab1_files&gt; is directory containing ab1 trace files of a single gene (pfcrt, pfmdr1, dhps, or dhfr)
    &lt;reference_gene_coding_sequence&gt; is path to reference coding sequence of respective gene
    &lt;gene&gt; is pfcrt, pfmdr1, dhps, or dhfr
    &lt;quality_cutoff&gt; is the error probability for trimming bases. 
                    Trimming is based on modified Richard Mott's algorithm.
                    This argument is OPTIONAL. Default value is 0.05.
                    Legal values range from 0 to 1.
</code></pre>

<hr>

<h2>
<a id="example-command" class="anchor" href="#example-command" aria-hidden="true"><span class="octicon octicon-link"></span></a>Example command</h2>

<pre><code>    bash anti_malaria_snps.sh ~/pfcrt_ab1_seq/ ~/anti_mdr_snps/pfcrt_pf3D7_cds.fasta pfcrt 0.01
</code></pre>

<hr>

<h2>
<a id="output" class="anchor" href="#output" aria-hidden="true"><span class="octicon octicon-link"></span></a>Output</h2>

<p>The output file is named in this format: <code>gene_dd_mm_yy.csv</code> and is stored in the <code>&lt;directory_of_ab1_files&gt;</code>
The output for the example above could look like this:</p>

<table>
<thead>
<tr>
<th>Sample</th>
<th align="center">Codon_72</th>
<th align="center">Codon_72_aa</th>
<th align="center">Codon_73</th>
<th align="center">Codon_73_aa</th>
<th align="center">Codon_74</th>
<th align="center">Codon_74_aa</th>
<th align="center">Codon_75</th>
<th align="center">Codon_75_aa</th>
<th align="center">Codon_76</th>
<th align="center">Codon_76_aa</th>
</tr>
</thead>
<tbody>
<tr>
<td>89C_CRT</td>
<td align="center">TGT</td>
<td align="center">C</td>
<td align="center">GTA</td>
<td align="center">V</td>
<td align="center">ATG</td>
<td align="center">M</td>
<td align="center">AAT</td>
<td align="center">N</td>
<td align="center">AAA</td>
<td align="center">K</td>
</tr>
<tr>
<td>90C_CRT</td>
<td align="center">TGT</td>
<td align="center">C</td>
<td align="center">GTA</td>
<td align="center">V</td>
<td align="center">ATG</td>
<td align="center">M</td>
<td align="center">AAT</td>
<td align="center">N</td>
<td align="center">AAA</td>
<td align="center">K</td>
</tr>
<tr>
<td>92C_CRT</td>
<td align="center">TGT</td>
<td align="center">C</td>
<td align="center">GTA</td>
<td align="center">V</td>
<td align="center">ATG</td>
<td align="center">M</td>
<td align="center">AAT</td>
<td align="center">N</td>
<td align="center">AAA</td>
<td align="center">K</td>
</tr>
<tr>
<td>93C_CRT</td>
<td align="center">TGT</td>
<td align="center">C</td>
<td align="center">GTA</td>
<td align="center">V</td>
<td align="center">ATG</td>
<td align="center">M</td>
<td align="center">AAT</td>
<td align="center">N</td>
<td align="center">AAA</td>
<td align="center">K</td>
</tr>
<tr>
<td>94C_CRT</td>
<td align="center">TGT</td>
<td align="center">C</td>
<td align="center">GTA</td>
<td align="center">V</td>
<td align="center">ATG</td>
<td align="center">M</td>
<td align="center">AAT</td>
<td align="center">N</td>
<td align="center">AAA</td>
<td align="center">K</td>
</tr>
<tr>
<td>97C_CRT</td>
<td align="center">TGT</td>
<td align="center">C</td>
<td align="center">GTA</td>
<td align="center">V</td>
<td align="center">ATG</td>
<td align="center">M</td>
<td align="center">AAT</td>
<td align="center">N</td>
<td align="center">AAA</td>
<td align="center">K</td>
</tr>
<tr>
<td>99C_CRT</td>
<td align="center">TGT</td>
<td align="center">C</td>
<td align="center">GTA</td>
<td align="center">V</td>
<td align="center">ATG</td>
<td align="center">M</td>
<td align="center">AAT</td>
<td align="center">N</td>
<td align="center">AAA</td>
<td align="center">K</td>
</tr>
<tr>
<td>PfCRT_ref</td>
<td align="center">TGT</td>
<td align="center">C</td>
<td align="center">GTA</td>
<td align="center">V</td>
<td align="center">ATG</td>
<td align="center">M</td>
<td align="center">AAT</td>
<td align="center">N</td>
<td align="center">AAA</td>
<td align="center">K</td>
</tr>
</tbody>
</table>

<hr>

<h2>
<a id="example-data" class="anchor" href="#example-data" aria-hidden="true"><span class="octicon octicon-link"></span></a>Example data</h2>

<p>The archived file containing the source code for this pipeline also contains some <code>example_data</code>.
The <code>example_data</code> directory contains:</p>

<ol>
<li><p><code>example_ab1</code>             : This directory contains 50 ab1 trace files obtained from sequencing regions of <em>pfcrt</em> 
                           spanning codons 72 - 76 from 50 <em>P. falciparum</em> isolates.
                           This directory can be used as <code>&lt;directory_of_ab1_files&gt;</code>. 
                           It also contains 3 intermediate files (<code>pfcrt_30_06_15.fasta</code>, <code>pfcrt_30_06_15.sam</code>, 
                           <code>pfcrt_30_06_15.fastq</code>) and the main output file (<code>pfcrt_30_06_15.csv</code>).
                           Typically, <code>&lt;directory_of_ab1_files&gt;</code> will not contain any intermediate files or 
                           output file until you have ran the pipeline successfully.</p></li>
<li><p><code>pfcrt_pf3d7_cds.fasta</code>   : The reference coding sequence for <em>pfcrt</em>. 
                           This file can be used as <code>&lt;reference_gene_coding_sequence&gt;</code>. 
                           Reference was originally downloaded from <a href="http://plasmodb.org/plasmo/">plasmodb.org</a>.</p></li>
</ol>
        </section>

        <aside id="sidebar">
          <a href="https://github.com/Setfelix/anti_malaria_snps/zipball/master" class="button">
            <small>Download</small>
            .zip file
          </a>
          <a href="https://github.com/Setfelix/anti_malaria_snps/tarball/master" class="button">
            <small>Download</small>
            .tar.gz file
          </a>

          <p class="repo-owner"><a href="https://github.com/Setfelix/anti_malaria_snps"></a> is maintained by <a href="https://github.com/Setfelix">Setfelix</a>.</p>

          <p>This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the Architect theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.</p>
        </aside>
      </div>
    </div>

  
  </body>
</html>
