<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>Anti-malarial SNPs by Setfelix</title>
  </head>

  <body>
    <header>
      <div class="inner">
        <h1>Anti-malarial SNPs</h1>
        <h2>Pipeline to detect SNPs in specific codons of genes associated with antimalarial drug resistance</h2>
        <a href="https://github.com/Setfelix/anti_malaria_snps" class="button"><small>View project on</small> GitHub</a>
      </div>
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
          <h1>
<a id="description" class="anchor" href="#description" aria-hidden="true"><span class="octicon octicon-link"></span></a>Description</h1>

<p><code>anti_malaria_snps.sh</code> is a BASH-scripted pipeline that can be used to identify SNPs in specific codons of 
<em>Plasmodium falciparum</em> genes that are associated with antimalarial drug resistance. 
Currently, the following genes and codons are supported <em>pfcrt, pfmdr1, dhps, dhfr</em>.</p>

<p>Using Sanger sequencing ab1 trace files of whole gene (or regions spanning codons of interest) from several samples, 
this pipeline outputs a CSV-formatted file of sample name, codons and their corresponding amino acid.</p>

<p>This pipeline is a combination of existing tools (such as <a href="http://bio-bwa.sourceforge.net/">BWA</a>, <a href="http://sourceforge.net/projects/sam2fasta/files/">sam2fasta.py</a>, <a href="https://github.com/bow/abifpy">abifpy</a>) and custom scripts and is made up of the following steps:</p>

<ol>
<li>Base calling and quality control, </li>
<li>Alignment of trimmed and filtered sequences to coding sequence of respective reference gene, </li>
<li>Conversion of alignment from SAM to FASTA format, </li>
<li>Parse FASTA alignment to output codons and corresponding amino acid in CSV format.</li>
</ol>

<hr>

<h2>
<a id="usage" class="anchor" href="#usage" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage</h2>

<pre><code>     bash anti_malaria_snps.sh &lt;directory_of_ab1_files&gt; &lt;reference_gene_coding_sequence&gt; &lt;gene&gt;
</code></pre>

<p>where: </p>

<pre><code>    &lt;directory_of_ab1_files&gt; is directory containing ab1 trace files of a single gene
    &lt;reference_gene_coding_sequence&gt; is path to reference coding sequence
    &lt;gene&gt; is pfcrt, pfmdr1, dhps, or dhfr
</code></pre>

<hr>

<h2>
<a id="example" class="anchor" href="#example" aria-hidden="true"><span class="octicon octicon-link"></span></a>Example</h2>

<pre><code>    bash anti_malaria_snps.sh ~/pfcrt_ab1_seq/ ~/anti_mdr_snps/pfcrt_pf3D7_cds.fasta pfcrt
</code></pre>

<hr>

<h2>
<a id="sample-output" class="anchor" href="#sample-output" aria-hidden="true"><span class="octicon octicon-link"></span></a>Sample output</h2>

<table>
<thead>
<tr>
<th>Sample</th>
<th>Codon_72</th>
<th>Codon_72_aa</th>
<th>Codon_73</th>
<th>Codon_73_aa</th>
<th>Codon_74</th>
<th>Codon_74_aa</th>
<th>Codon_75</th>
<th>Codon_75_aa</th>
<th>Codon_76</th>
<th>Codon_76_aa</th>
</tr>
</thead>
<tbody>
<tr>
<td>89C_CRT</td>
<td>TGT</td>
<td>C</td>
<td>GTA</td>
<td>V</td>
<td>ATG</td>
<td>M</td>
<td>AAT</td>
<td>N</td>
<td>AAA</td>
<td>K</td>
</tr>
<tr>
<td>90C_CRT</td>
<td>TGT</td>
<td>C</td>
<td>GTA</td>
<td>V</td>
<td>ATG</td>
<td>M</td>
<td>AAT</td>
<td>N</td>
<td>AAA</td>
<td>K</td>
</tr>
<tr>
<td>92C_CRT</td>
<td>TGT</td>
<td>C</td>
<td>GTA</td>
<td>V</td>
<td>ATG</td>
<td>M</td>
<td>AAT</td>
<td>N</td>
<td>AAA</td>
<td>K</td>
</tr>
<tr>
<td>93C_CRT</td>
<td>TGT</td>
<td>C</td>
<td>GTA</td>
<td>V</td>
<td>ATG</td>
<td>M</td>
<td>AAT</td>
<td>N</td>
<td>AAA</td>
<td>K</td>
</tr>
<tr>
<td>94C_CRT</td>
<td>TGT</td>
<td>C</td>
<td>GTA</td>
<td>V</td>
<td>ATG</td>
<td>M</td>
<td>AAT</td>
<td>N</td>
<td>AAA</td>
<td>K</td>
</tr>
<tr>
<td>97C_CRT</td>
<td>TGT</td>
<td>C</td>
<td>GTA</td>
<td>V</td>
<td>ATG</td>
<td>M</td>
<td>AAT</td>
<td>N</td>
<td>AAA</td>
<td>K</td>
</tr>
<tr>
<td>99C_CRT</td>
<td>TGT</td>
<td>C</td>
<td>GTA</td>
<td>V</td>
<td>ATG</td>
<td>M</td>
<td>AAT</td>
<td>N</td>
<td>AAA</td>
<td>K</td>
</tr>
<tr>
<td>PfCRT_ref</td>
<td>TGT</td>
<td>C</td>
<td>GTA</td>
<td>V</td>
<td>ATG</td>
<td>M</td>
<td>AAT</td>
<td>N</td>
<td>AAA</td>
<td>K</td>
</tr>
</tbody>
</table>
        </section>

        <aside id="sidebar">
          <a href="https://github.com/Setfelix/anti_malaria_snps/zipball/master" class="button">
            <small>Download</small>
            .zip file
          </a>
          <a href="https://github.com/Setfelix/anti_malaria_snps/tarball/master" class="button">
            <small>Download</small>
            .tar.gz file
          </a>

          <p class="repo-owner"><a href="https://github.com/Setfelix/anti_malaria_snps"></a> is maintained by <a href="https://github.com/Setfelix">Setfelix</a>.</p>

          <p>This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the Architect theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.</p>
        </aside>
      </div>
    </div>

  
  </body>
</html>
